addons/account/tests/test_account_move_send.py

```diff
@@ -689,6 +691,180 @@ def test_compute_value_of_send_invoice_batch_wizard(self):
+            'res_id': invoice.id,
+            'is_note': False,
+            'scheduled_date': self.reference_now + timedelta(days=1),
+        }])
+
+        # Sending invoice now
+        scheduled_message._post_message()
+
+        # The PDF has been successfully generated.
+        pdf_report = invoice.invoice_pdf_report_id
+        self.assertTrue(pdf_report)
+        invoice_attachments = self.env['ir.attachment'].search([
+            ('res_model', '=', invoice._name),
+            ('res_id', '=', invoice.id),
+            ('res_field', '=', 'invoice_pdf_report_file'),
+        ])
+        self.assertEqual(len(invoice_attachments), 1)
+        self.assertTrue(self._get_mail_message(invoice))
+
+        # Send it again. The PDF must not be created again.
+        wizard = self.create_send_and_print(invoice, sending_methods=['email', 'manual'])
+        wizard.write({'scheduled_date': FieldDatetime.to_string(self.reference_now + timedelta(days=1))})
+        with patch('odoo.addons.account.models.account_move_send.AccountMoveSend._hook_invoice_document_after_pdf_report_render') as mocked_method:
+            with self.mock_datetime_and_now(self.reference_now):
+                wizard.action_schedule_message()
+            mocked_method.assert_not_called()
+        self.assertFalse(invoice.sending_data)
+
+        scheduled_message = self.env['mail.scheduled.message'].search([
+            ['model', '=', invoice._name],
+            ['res_id', '=', invoice.id],
+        ])
+        scheduled_message._post_message()
+
+        self.assertRecordValues(invoice, [{'invoice_pdf_report_id': pdf_report.id}])
+        invoice_attachments = self.env['ir.attachment'].search([
+            ('res_model', '=', invoice._name),
+            ('res_id', '=', invoice.id),
+            ('res_field', '=', 'invoice_pdf_report_file'),
+        ])
+        self.assertEqual(len(invoice_attachments), 1)
+        self.assertTrue(self._get_mail_message(invoice))
+
+        # Only one PDF linked to the invoice.
+        invoice_attachments = self.env['ir.attachment'].search([
+            ('name', '=', pdf_report.name),
+            ('res_model', '=', invoice._name),
+            ('res_id', '=', invoice.id),
+            ('res_field', '=', False),
+        ])
+        self.assertFalse(invoice_attachments)
+
+    def test_invoice_scheduled_messages(self):
+        partner_vals = {
+            'name': "Partner with address",
+            'phone': "+91 0123456789",
+            'email': "info@odoo.example.com",
+            'street': "Tower-3, Infocity",
+            'zip': "382010",
+            'city': "Gandhinagar",
+            'country_id': self.env.ref('base.in').id,
+        }
+        partner_with_address = self.env['res.partner'].create(partner_vals)
+        invoice = self.init_invoice("out_invoice", partner=partner_with_address, amounts=[1000], post=True)
+        wizard = self.create_send_and_print(invoice, sending_methods=['snailmail'])
```

Comment:
Should probably be in snailmail_account right?
(maybe some of the code in mail_scheduled_message override too?)